### DISCUSSION TITLE
Is State cloned for each request?

### DISCUSSION THREAD

**Discussion by soasada** [tokio-rs/axum#1911] [Q&A] [answered]
Hi, I'm a little bit confused with clone and Arc when using `with_state` function. 

I have the following AppState:

```rust
use tokio::sync::mpsc;

#[derive(Clone)]
pub struct ActorHandle {
    sender: mpsc::Sender<ActorMessage>,
}

impl ActorHandle {
  pub async fn send_msg(&self, msg: ActorMessage) {
    self.sender.send(msg).await;
  }
}

#[derive(Clone)]
pub struct AppState {
    pub actor_handle: ActorHandle,
}
```

When I create my axum router I create a regular AppState without Arc:

```rust
use tokio::sync::mpsc;

let (sender, receiver) = mpsc::channel(32);

let app_state = AppState {
        actor_handle: ActorHandle { sender }
};

let router = Router::new()
  .route("/foo", post(foo))
  .with_state(app_state);
```

Then I use it in the handler as a regular variable:

```rust
pub async fn foo(State(state): State<AppState>) -> impl IntoResponse {
  // msg creation omitted for brevity 
  state.actor_handle.send_msg(msg).await; // should I use state.clone() and then use the variable "actor_handle"?
}
```

Is this legal? do I need to clone or add Arc type to my state if this is my use case?

**Comment by davidpdrsn** [comment, marked as answer]
Yes the state is cloned for each request. Otherwise you wouldn't be able to extract an owned state.

**Reply by soasada** [reply]
Then, when are the use cases of using Arc wrapping the State?

**Reply by davidpdrsn** [reply]
If you have a state that is expensive to clone or can't be cloned then you need an `Arc`.

**Reply by thomasmost** [reply]
is this recommended for database pools, for example?

**Reply by davidpdrsn** [reply]
No. Pools are often arcs internally.

**Reply by thomasmost** [reply]
Got it! Thank you so much

**Reply by musjj** [reply]
If you have substates like:

```rust
struct Foo;
struct Bar;

struct AppState {
    foo: Foo,
    bar: Bar,
}

impl FromRef<AppState> for Foo {
    fn from_ref(input: &AppState) -> Self {
        input.foo.clone()
    }
}

fn my_route(
    State(foo): State<Foo>,
) {
    todo!()
}
```

Is `AppState` cloned on every request too? Or just `Foo`?

**Reply by mladedav** [reply]
It's cloned on every request. Or rather whatever you used in the `with_state` call for the give route that the request matched will be used.

**Reply by musjj** [reply]
So with my example, if you do:

```rust
let router = Router::new()
    .route("/my_route", my_route)
    .with_state(AppState { .. })
```

`AppState` will be cloned, then `Foo` will be cloned again afterwards?
